### PR TITLE
Remove interior mutability from TraitDef by turning fields into queries

### PR DESCRIPTION
**Pull request by michaelwoerister**
This PR gets rid of anything `std::cell` in `TraitDef` by 
- moving the global list of trait impls from `TraitDef` into a query,
- moving the list of trait impls relevent for some self-type from `TraitDef` into a query
- moving the specialization graph of trait impls into a query, and
- moving `TraitDef::object_safety` into a query.

I really like how querifying things not only helps with incremental compilation and on-demand, but also just plain makes the code cleaner `:)`

There are also some smaller fixes in the PR. Commits can be reviewed separately.

r? @eddyb or @nikomatsakis

### CODE REVIEW

File: src/librustc/traits/mod.rs [hunk 1]
```diff
@@ -782,3 +780,19 @@ impl<'tcx> TraitObligation<'tcx> {
...
+
+pub fn provide(providers: &mut ty::maps::Providers) {
+    *providers = ty::maps::Providers {
+        is_object_safe: object_safety::is_object_safe_provider,
+        specialization_graph_of: specialize::specialization_graph_provider,
+        ..*providers
+    };
+}
+
+pub fn provide_extern(providers: &mut ty::maps::Providers) {
```
Comment (eddyb) [inline src/librustc/traits/mod.rs]: Since they're identical I'd use the same function twice.

File: src/librustc/traits/project.rs [hunk 2]
```diff
@@ -1320,23 +1320,10 @@ fn assoc_ty_def<'cx, 'gcx, 'tcx>(
     let trait_def_id = selcx.tcx().impl_trait_ref(impl_def_id).unwrap().def_id;
     let trait_def = selcx.tcx().trait_def(trait_def_id);
 
-    if !trait_def.is_complete(selcx.tcx()) {
```
Comment (eddyb) [inline src/librustc/traits/project.rs]: How does this work? Do we produce cycle errors in more cases?

File: src/librustc/ty/maps.rs [hunk 3]
```diff
@@ -820,6 +849,13 @@ define_maps! { <'tcx>
     [] item_body_nested_bodies: ItemBodyNestedBodies(DefId) -> Rc<BTreeMap<hir::BodyId, hir::Body>>,
     [] const_is_rvalue_promotable_to_static: ConstIsRvaluePromotableToStatic(DefId) -> bool,
     [] is_mir_available: IsMirAvailable(DefId) -> bool,
+
+    [] trait_impls_of: TraitImpls(DefId) -> Rc<Vec<DefId>>,
+    // Note that TraitDef::for_each_relevant_impl() will do type simplication for you.
+    [] relevant_trait_impls_for: relevant_trait_impls_for((DefId, SimplifiedType))
+        -> Rc<Vec<DefId>>,
```
Comment (eddyb) [inline src/librustc/ty/maps.rs]: Wait, where are all the blanket impls?

File: src/librustc/ty/trait_def.rs [hunk 4]
```diff
@@ -304,29 +82,71 @@ impl<'a, 'gcx, 'tcx> TraitDef {
...
+        .filter(|&impl_def_id| {
+            let impl_trait_ref = tcx.impl_trait_ref(impl_def_id).unwrap();
+            let impl_simple_self_ty = fast_reject::simplify_type(tcx,
+                                                                 impl_trait_ref.self_ty(),
+                                                                 false);
+            if let Some(impl_simple_self_ty) = impl_simple_self_ty {
+                impl_simple_self_ty == self_ty
+            } else {
+                // blanket impl (?)
+                true
```
Comment (eddyb) [inline src/librustc/ty/trait_def.rs]: This is potentially very bad, as N blanket impls and M non-blanket now use N*M space.

**Review by eddyb — COMMENTED** [review]
I like the overall approach but have some doubts on the details.

File: src/librustc/ty/trait_def.rs (hunk 4 above)
Comment (michaelwoerister) [inline src/librustc/ty/trait_def.rs]: Hm, OK, let me think about that. Maybe we could share the list of blanket impls?

File: src/librustc/ty/maps.rs (hunk 3 above)
Comment (michaelwoerister) [inline src/librustc/ty/maps.rs]: They are part of the list.

File: src/librustc/traits/project.rs (hunk 2 above)
Comment (michaelwoerister) [inline src/librustc/traits/project.rs]: I'm not sure what you mean exactly. I don't see a potential cycle here.

File: src/librustc/traits/mod.rs (hunk 1 above)
Comment (michaelwoerister) [inline src/librustc/traits/mod.rs]: You think so? It's just a coincidence that they are identical at the moment.

Comment (nikomatsakis) [inline src/librustc/traits/mod.rs]: maybe have `provide()` call `provide_extern()`? I imagine that anything we provide for extern crates, we would also provide for locals, but  maybe not the other way around

File: src/librustc/traits/project.rs (hunk 2 above)
Comment (eddyb) [inline src/librustc/traits/project.rs]: The `!is_complete` case is a fallback for cycles, that is, when normalization of a projection like `<T as Foo>::Assoc` is required to compute `Foo`'s specialization graph.

Comment (eddyb) [inline src/librustc/traits/project.rs]: cc @aturon We may be lacking tests here.

Comment (michaelwoerister) [inline src/librustc/traits/project.rs]: Oh, so `assoc_ty_def()` might be called while still building the specialization graph?

File: src/librustc/traits/mod.rs (hunk 1 above)
Comment (eddyb) [inline src/librustc/traits/mod.rs]: That makes sense.

File: src/librustc/traits/project.rs (hunk 2 above)
Comment (eddyb) [inline src/librustc/traits/project.rs]: The system was certainly designed with that in mind, yeah.

Comment (eddyb) [inline src/librustc/traits/project.rs]: See #32205 and [the docs for `ExactMatch`](https://github.com/rust-lang/rust/blob/9890e0466d951d1b0020ae38e2a4c0d8c43cfea1/src/librustc/traits/project.rs#L41-L69) (now controlled by `!is_complete`).

Comment (michaelwoerister) [inline src/librustc/traits/project.rs]: Thanks, @eddyb! 
How does the following look to you?
```rust
fn assoc_ty_def<'cx, 'gcx, 'tcx>(
    selcx: &SelectionContext<'cx, 'gcx, 'tcx>,
    impl_def_id: DefId,
    assoc_ty_name: ast::Name)
    -> Option<specialization_graph::NodeItem<ty::AssociatedItem>>
{
    let tcx = selcx.tcx();
    let trait_def_id = tcx.impl_trait_ref(impl_def_id).unwrap().def_id;
    let trait_def = tcx.trait_def(trait_def_id);

    // Check for a cycle
    let specialization_graph_complete =
        ty::queries::specialization_graph_of::try_get(tcx, DUMMY_SP, self.def_id).is_ok();

    if !specialization_graph_complete {
        let impl_node = specialization_graph::Node::Impl(impl_def_id);
        for item in impl_node.items(selcx.tcx()) {
            if item.kind == ty::AssociatedKind::Type && item.name == assoc_ty_name {
                return Some(specialization_graph::NodeItem {
                    node: specialization_graph::Node::Impl(impl_def_id),
                    item: item,
                });
            }
        }
        None
    } else {
        trait_def
            .ancestors(tcx, impl_def_id)
            .defs(tcx, assoc_ty_name, ty::AssociatedKind::Type)
            .next()
    }
}
```
(I would clean this up of course and add some comments)

Comment (eddyb) [inline src/librustc/traits/project.rs]: That's probably fine, yeah, I expect it to be more or less equivalent to `is_complete`.
If you have a build without that change, can you try to synthesize a cycle for a test?

Comment (michaelwoerister) [inline src/librustc/traits/project.rs]: I'm having a hard time producing a crash. The following example seems to compile just fine:
```
#![feature(specialization)]

trait Assoc {
    type Output;
}

impl<T> Assoc for T {
    type Output = bool;
}

impl Assoc for u8 {} // <- inherits the non-default type from above
trait Foo {}
impl Foo for u32 {}
impl Foo for <u8 as Assoc>::Output {} // <- this projection will fail

fn main() {
    let _: &Foo = &false;
}
```

Comment (eddyb) [inline src/librustc/traits/project.rs]: It needs to be used *in the specialization graph of the same trait*.

Comment (michaelwoerister) [inline src/librustc/traits/project.rs]: > It needs to be used in the specialization graph of the same trait.

Yes, that makes sense. And I thought the following example might trigger the condition, but it doesn't `:(`
```rust
#![feature(specialization)]

trait Assoc {
    type Output;
}

impl<T> Assoc for T {
    default type Output = u8;
}

impl Assoc for u32 {
    type Output = u32;
}

impl Assoc for u64 {
    // Shouldn't this cause a reentrant call?
    type Output = <u32 as Assoc>::Output;
}

fn main() {
    let _x: <u32 as Assoc>::Output = 0;
}
```

Comment (eddyb) [inline src/librustc/traits/project.rs]: In the *impl header* - only those are checked by coherence/specialization.

Comment (michaelwoerister) [inline src/librustc/traits/project.rs]: So, the following gives me the error `overflow evaluating the requirement <u32 as Assoc>::Output` regardless of whether cycle detection is enabled:

```rust

#![feature(specialization)]

trait Assoc {
    type Output;
}

impl Assoc for u32 {
    type Output = u64;
}

impl Assoc for <u32 as Assoc>::Output {
    type Output = u64;
}

fn main() {
    let _x: <u64 as Assoc>::Output = 0;
}
```
So that's not really what we are looking for, I guess?

**Comment by michaelwoerister** [comment]
@eddyb I added the cycle check back in (https://github.com/rust-lang/rust/pull/41911/commits/40a6734ae1a1ec9ad96b8f5b58a64d10d0cdbe76) and blanket impl lists are now shared (https://github.com/rust-lang/rust/pull/41911/commits/742ebc17ff4d033f6b5c6348ddb1c94d730c56b8).

But I didn't manage to come up with a test case that would cause a cycle (see https://github.com/rust-lang/rust/pull/41911#discussion_r116262605), unfortunately.

**Comment by eddyb** [comment]
@michaelwoerister Yeah, I'm waiting for @aturon and/or @nikomatsakis for that.

File: src/librustc/hir/map/mod.rs [hunk 5]
```diff
@@ -497,15 +497,15 @@ impl<'hir> Map<'hir> {
     }
 
     pub fn trait_impls(&self, trait_did: DefId) -> &'hir [NodeId] {
-        self.dep_graph.read(DepNode::TraitImpls(trait_did));
+        self.dep_graph.read(DepNode::AllLocalTraitImpls);
```
Comment (nikomatsakis) [inline src/librustc/hir/map/mod.rs]: I'm curious why you converted this into a single dep-node, rather than keeping it per-trait. It seems like this is losing quite a bit of precision, no? 

(That is, if **any** trait adds an impl, we will consider them **all** to have changed.)

File: src/librustc/traits/project.rs (hunk 2 above)
Comment (nikomatsakis) [inline src/librustc/traits/project.rs]: Hmm, I don't really expect you to be able to project `<u32 as Assoc>::Output` in an impl for `Assoc`. What I expect *could* work is if there is a DAG between the traits, sort of like this:

```rust
#![feature(specialization)]

use std::vec;

trait Assoc {
    type Output;
}

impl Assoc for u32 {
    type Output = u64;
}

impl Assoc for <vec::IntoIter<u64> as Iterator>::Item {
    type Output = u64;
}

fn main() {
}
```

Interestingly, coherence rejects this example, presumably because of RFC 1214 concerns. I'm not sure that's right, but this example does work today:

```rust
#![feature(specialization)]

use std::vec;

trait Assoc {
    type Output;
}

impl Assoc for u32 {
    type Output = u64;
}

impl Assoc for <u32 as Assoc2>::Output {
    type Output = u64;
}

trait Assoc2 {
    type Output;
}

impl Assoc2 for u32 {
    type Output = u64;
}

fn main() {
}
```

That said, I think that would also work in @michaelwoerister's branch, no?

File: src/librustc/hir/map/mod.rs (hunk 5 above)
Comment (michaelwoerister) [inline src/librustc/hir/map/mod.rs]: Mostly for ease of implementation and symmetry with the situation in metadata. `DepNode::TraitImpls` has been re-purposed to represent local *and* remote impls and, with red-green, will take care of stopping invalidation short.

Without red-green though you are right, it's losing precision. Want me to make it per-item?

File: src/librustc/ty/trait_def.rs [hunk 6]
```diff
@@ -304,29 +136,86 @@ impl<'a, 'gcx, 'tcx> TraitDef {
...
+
+    let relevant: Vec<DefId> = all_trait_impls
+        .non_blanket_impls
+        .iter()
+        .cloned()
+        .filter(|&impl_def_id| {
+            let impl_trait_ref = tcx.impl_trait_ref(impl_def_id).unwrap();
+            let impl_simple_self_ty = fast_reject::simplify_type(tcx,
+                                                                 impl_trait_ref.self_ty(),
+                                                                 false).unwrap();
```
Comment (eddyb) [inline src/librustc/ty/trait_def.rs]: I just realized, this is stricter than it has to be. it should use `tcx.type_of(impl_def_id)` instead.

File: src/librustc/ty/trait_def.rs [hunk 7]
```diff
@@ -304,29 +136,86 @@ impl<'a, 'gcx, 'tcx> TraitDef {
...
+    let mut blanket_impls = Vec::new();
+    let mut non_blanket_impls = Vec::new();
+
+    let local_impls = tcx.hir
+                         .trait_impls(trait_id)
+                         .into_iter()
+                         .map(|&node_id| tcx.hir.local_def_id(node_id));
+
+     for impl_def_id in local_impls.chain(remote_impls.into_iter()) {
+        let impl_trait_ref = tcx.impl_trait_ref(impl_def_id).unwrap();
```
Comment (eddyb) [inline src/librustc/ty/trait_def.rs]: Same here, I suppose.

File: src/test/compile-fail/coherence-inherited-assoc-ty-cycle-err.rs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Formerly this ICEd with the following message:
+// Tried to project an inherited associated type during coherence checking,
+// which is currently not supported.
```
Comment (eddyb) [inline src/test/compile-fail/coherence-inherited-assoc-ty-cycle-err.rs]: Where is that message in the source? Could that codepath be removed?

Comment (michaelwoerister) [inline src/test/compile-fail/coherence-inherited-assoc-ty-cycle-err.rs]: It here: https://github.com/rust-lang/rust/blob/4d09a0eb5d201c831d19f3fe93fbe636eca7d97d/src/librustc/traits/project.rs#L989

But it has this nice, long comment which made me reluctant to remove it. I could move the comment to `assoc_ty_def()`, make its return value non-optional, and then remove the branch though.

Comment (eddyb) [inline src/test/compile-fail/coherence-inherited-assoc-ty-cycle-err.rs]: Yeah, moving the comment should be fine.

**Comment by michaelwoerister** [comment]
@eddyb I'm not entirely sure if that last commit (08660af) is an improvement.

File: src/librustc/hir/map/mod.rs (hunk 5 above)
Comment (nikomatsakis) [inline src/librustc/hir/map/mod.rs]: Probably not a big deal.

File: src/librustc/traits/project.rs [hunk 9]
```diff
@@ -900,96 +900,50 @@ fn assemble_candidates_from_impls<'cx, 'gcx, 'tcx>(
...
-                    //
-                    // The essential problem here is that the
-                    // projection fails, leaving two unnormalized
-                    // types, which appear not to unify -- so the
-                    // overlap check succeeds, when it should
-                    // fail.
-                    span_bug!(obligation.cause.span,
-                              "Tried to project an inherited associated type during \
-                               coherence checking, which is currently not supported.");
+                    None
```
Comment (nikomatsakis) [inline src/librustc/traits/project.rs]: Why did you remove this comment etc? Do we think this code-path is dead, or not?

Comment (nikomatsakis) [inline src/librustc/traits/project.rs]: If we think it's unreachable, I'd rather see `bug!(...)`, probably with an updated comment explaining that we **expect** this scenario to be prevented by cycle-errors.

Comment (michaelwoerister) [inline src/librustc/traits/project.rs]: The diff is a bit confusion here because indentation has changed. The `None` belongs to `let new_candidate = if !is_default {`, for which nothing should have changed, I believe.

Comment (nikomatsakis) [inline src/librustc/traits/project.rs]: 👍 I see now
